packages/ses/src/scope-terminator.js

```diff
@@ -0,0 +1,103 @@
+import {
+  Proxy,
+  String,
+  TypeError,
+  create,
+  freeze,
+  getOwnPropertyDescriptors,
+  globalThis,
+  immutableObject,
+  reflectSet,
+} from './commons.js';
+import { assert } from './error/assert.js';
+
+const { details: d, quote: q } = assert;
+
+/**
+ * alwaysThrowHandler
+ * This is an object that throws if any property is called. It's used as
+ * a proxy handler which throws on any trap called.
+ * It's made from a proxy with a get trap that throws. It's safe to
+ * create one and share it between all Proxy handlers.
+ */
+const alwaysThrowHandler = new Proxy(
+  immutableObject,
+  freeze({
+    get(_shadow, prop) {
+      // eslint-disable-next-line @endo/no-polymorphic-call
+      assert.fail(
+        d`Please report unexpected scope handler trap: ${q(String(prop))}`,
+      );
+    },
+  }),
+);
+
+/*
+ * createScopeTerminator()
+ * scopeTerminatorHandler manages a scopeTerminator Proxy which serves as
+ * the final scope boundary that will always return "undefined" in order
+ * to prevent access to "start compartment globals". When "sloppyGlobalsMode"
+ * is true, the Proxy will perform sets on the "globalObject".
+ */
+export const createScopeTerminator = (
+  globalObject,
+  { sloppyGlobalsMode = false } = {},
+) => {
+  const scopeProxyHandlerProperties = {
+    get(_shadow, _prop) {
+      return undefined;
+    },
+
+    set(_shadow, prop, value) {
+      if (sloppyGlobalsMode) {
+        // Redirect set properties to the globalObject.
+        return reflectSet(globalObject, prop, value);
+      } else {
+        return false;
+      }
```

Comment:
The fact this check didn't exist before seems like a bug to me, but I don't have enough historical background.

Reply:
Yeah, same suspicion

Reply:
ok so its about the combination of the `has` and `set` implementation.

if `has` returns false, it will not hit the `set` hook and instead just naturally throw a `ReferenceError` (if it doesnt exist on realm globalThis).

if `has` returns true then we hit `set` where we [previously](https://github.com/endojs/endo/blob/705aef24f34bb9794f0aa807d567b3efbf0c23af/packages/ses/src/scope-handler.js#L128) just forwarded the set to the `globalObject`. in this refactor, we only hit `scopeTerminator` set hook if the property is missing from the globalObject at which point we should deny the set / throw a ReferenceError (strict mode) or redirect the set to the globalObject (sloppyGlobals)

Reply:
added a test for this and recorded in pr top comment